File: neetcode all and misc questions/Arrays and Strings/encode_and_decode_tinyurl.py
"""
Key Concept:

I'm using Python's hash function such that
we take a given long URL and turn it into the same short URL every time,
so we don't need to worry about using random chars and checking to make sure that sequence of random chars
hasn't been used.

Just store the hashed version as the shortURL in a dict, mapping to the original long form URL,
and then retrieve that in decode 

If the problem stated that there was a character limitation for the amount of characters in the short URL,
we'd need to change this solution to use random chars instead, and then within our encode,
we need to check if that random sequence of char hasn't been used already.

"""
class Codec:
    def __init__(self):
        self.decoder = dict()
        self.baseUrl = "http://tinyurl.com/"
    def encode(self, longUrl: str) -> str:
        """Encodes a URL to a shortened URL.
        """
        shortUrl = self.baseUrl + str(hash(longUrl))
        
        self.decoder[shortUrl] = longUrl
        return shortUrl
    def decode(self, shortUrl: str) -> str:
        """Decodes a shortened URL to its original URL.
        """
        return self.decoder[shortUrl]
        

# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(url))

File: node_js_ds_and_algos_implementations/arrays and strings/hashmaps/encode_and_decode_tinyurl.js
/**
 * Encodes a URL to a shortened URL.
 *
 * @param {string} longUrl
 * @return {string}
 */

// keep forwards and backwards mapping
// use an autoincrement ID as the short URL
// note that in real interview settings, they may ask for limitations,
// like a 10 character max tiny URL. In this case, using only digits would allow up to
// 9^10 possibilities, but including characters would raise the limit to 62^10

var encodeMap = {}
var decodeMap = {}
var total = 0

var encode = function(longUrl) {
    if (!(longUrl in encodeMap)){
        total++
        encodeMap[longUrl] = total
        decodeMap[total] = longUrl
        return total
    }    
};

/**
 * Decodes a shortened URL to its original URL.
 *
 * @param {string} shortUrl
 * @return {string}
 */
var decode = function(shortUrl) {
    if (shortUrl in decodeMap){
        return decodeMap[shortUrl]
    }
};

/**
 * Your functions will be called as such:
 * decode(encode(url));
 */